lightning-invoice/src/utils.rs

```diff
@@ -8,19 +8,115 @@ use bitcoin_hashes::Hash;
 use lightning::ln::{PaymentHash, PaymentPreimage, PaymentSecret};
-use lightning::ln::channelmanager::{ChannelDetails, ChannelManager, PaymentId, PaymentSendFailure, MIN_FINAL_CLTV_EXPIRY};
+use lightning::ln::channelmanager::{ChannelDetails, ChannelManager, PaymentId, PaymentSendFailure, MIN_FINAL_CLTV_EXPIRY, MIN_CLTV_EXPIRY_DELTA};
 use lightning::ln::msgs::LightningError;
 use lightning::routing::scoring::Score;
 use lightning::routing::network_graph::{NetworkGraph, RoutingFees};
 use lightning::routing::router::{Route, RouteHint, RouteHintHop, RouteParameters, find_route};
 use lightning::util::logger::Logger;
+use secp256k1::Secp256k1;
 use secp256k1::key::PublicKey;
 use core::convert::TryInto;
 use core::ops::Deref;
 use core::time::Duration;
 
+#[cfg(feature = "std")]
+/// Utility to create an invoice that can be paid to one of multiple nodes, or a "phantom invoice."
+/// This works because the invoice officially pays to a fake node (the "phantom"). Useful for
+/// load-balancing payments between multiple nodes.
+///
+/// `channels` parameter:
+/// * Contains a list of tuples of `(phantom_node_scid, real_node_pubkey,
+///   real_channel_details)`, where `phantom_node_scid` is unique to each `channels` entry
+/// * `phantom_node_scid`s are retrieved from [`ChannelManager::get_phantom_scid`]
+/// * `real_channel_details` are retrieved from [`ChannelManager::list_channels`]
+/// * Contains channel info for all nodes participating in the phantom invoice
+/// * It is fine to cache `channels` and reuse it across invoices, as long as the data is
+///   updated when a channel becomes disabled or closes
+/// * Note that including too many channels here could result in making the invoice too long for QR
+///   code scanning
+///
+/// `payment_hash` and `payment_secret` come from [`ChannelManager::create_inbound_payment`] or
+/// [`ChannelManager::create_inbound_payment_for_hash`].
+///
+/// See [`KeysInterface::get_phantom_secret`] for more requirements on supporting phantom node
+/// payments.
+///
+/// [`KeysInterface::get_phantom_secret`]: lightning::chain::keysinterface::KeysInterface::get_phantom_secret
+pub fn create_phantom_invoice<Signer: Sign, K: Deref>(
+	amt_msat: Option<u64>, description: String, payment_hash: PaymentHash, payment_secret:
+	PaymentSecret, channels: &[(u64, PublicKey, &ChannelDetails)], keys_manager: K, network: Currency
+) -> Result<Invoice, SignOrCreationError<()>> where K::Target: KeysInterface {
+	if channels.len() == 0 {
+		return Err(SignOrCreationError::CreationError(CreationError::MissingRouteHints))
+	}
+	let phantom_secret = match keys_manager.get_phantom_secret() {
+		Ok(s) => s,
+		Err(()) => return Err(SignOrCreationError::CreationError(CreationError::InvalidPhantomScid))
+	};
+	let phantom_pubkey = PublicKey::from_secret_key(&Secp256k1::new(), &phantom_secret);
+	let mut invoice = InvoiceBuilder::new(network)
+		.description(description)
+		.current_timestamp()
+		.payee_pub_key(phantom_pubkey)
```

Comment:
You shouldn't need this if the signature is a recoverable signature (which we require). That should let you drop the `get_phantom_secret` call entirely, I think?